crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -627,6 +632,89 @@ impl AppService {
         webserver::warp_filter(self.clone())
     }
 
+    /// Receive an incoming [transaction], pushing the contained events to
+    /// active virtual clients.
+    ///
+    /// [transaction]: https://spec.matrix.org/v1.2/application-service-api/#put_matrixappv1transactionstxnid
+    pub async fn receive_transaction(
+        &self,
+        transaction: push_events::v1::IncomingRequest,
+    ) -> Result<()> {
+        let client = self.get_cached_client(None)?;
+
+        // Find membership events affecting members in our namespace, and update
+        // membership accordingly
+        for event in transaction.events.iter() {
+            let event = match event.deserialize() {
+                Ok(AnyRoomEvent::State(AnyStateEvent::RoomMember(event))) => event,
+                _ => continue,
+            };
+            if !self.user_id_is_in_namespace(event.state_key())? {
+                continue;
+            }
+            let localpart = event.state_key().localpart();
+            client
+                .store()
+                .set_custom_value(
+                    &[USER_MEMBER, event.room_id().as_bytes(), localpart.as_bytes(), b"."].concat(),
+                    event.membership().to_string().into_bytes(),
+                )
+                .await?;
+        }
+
+        #[derive(Debug, Deserialize)]
+        struct EventRoomId {
+            room_id: Option<OwnedRoomId>,
+        }
+
+        // Send a sync event to each active client
+        // TODO: What do we do with state events?
```

Comment:
Could you expand on this? What is the exact problem with state events?

Reply:
I was uncertain if putting any state events received in the transaction into the `state` section of the synthesized sync event would be needed for the virtual clients to maintain a coherent view of the rooms they're participating in. Though the `state` section is for events from *before* the bits in `timeline`, so that adds an additional layer of complexity. I'm a bit unclear on what the right move is here.

Reply:
Since the major difference between what's in `state` vs what's in the `timeline` is the former will not be displayed to the user while the later will, and the app-service doesn't display any of this to users, we can safely just put everything into the timeline.

We just need to make sure that, what we would otherwise put into the `state` section, comes before anything that belongs into the `timeline` section.

I think this is already the case?

Reply:
I believe so, yes. In that case I don't think there's any need to change anything here. Thanks for helping me think through this.

Reply:
Let's just remove the TODO if there's nothing to do here. :)